### PR TITLE
Add flash attention for `gpt_bigcode`

### PR DESCRIPTION
**Pull request by susnato**
# What does this PR do?

Adds Flash Attention 2 for `GPTBigCode (Starcoder)` as discussed in in this issue - https://github.com/huggingface/transformers/issues/26350 .

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc : @younesbelkada

### CODE REVIEW

**Comment by susnato** [comment]
All tests are all passing except the `test_flash_attn_2_generate_use_cache` tests(same as opt). 
Screenshot from 2023-09-29 02-04-00

At this point, I think the failing of the tests might be caused due to some unknown error on my machine rather than the code since I have this test failing for all `flash-attention` supported models.

If you don't mind @younesbelkada , could you please checkout this branch and run the tests on your end and let me know the results?

**Comment by younesbelkada** [comment]
Again, thanks a lot for your amazing contribution @susnato !
Yes OK will check that tomorrow and let you know

**Comment by susnato** [comment]
Thanks a lot @younesbelkada! 
BTW if the tests pass on your machine then will it be ready to merge? Or we will need to investigate more why it fails on my end?

File: docs/source/en/perf_infer_gpu_one.md
Comment (younesbelkada) [inline docs/source/en/perf_infer_gpu_one.md:36]:
```suggestion
- GPTBigCode (Starcoder)
```

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py
Comment (younesbelkada) [inline src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:374]: This can be fixed in the global fix I want to apply in https://github.com/huggingface/transformers/issues/26451 as a follow up PR that I will take care

**Review by younesbelkada — APPROVED** [review]
LGTM I can confirm FA tests pass and I have also tried a generation locally and it seems to work great! Thanks for your huge work ! 🙏

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py
Comment (susnato) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:374]: should I then remove this block? or are we keeping this block for now.

Comment (younesbelkada) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:374]: I would say we can keep it for now

Comment (susnato) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:374]: ok

File: docs/source/en/perf_infer_gpu_one.md
Comment (susnato) [reply docs/source/en/perf_infer_gpu_one.md]: done

**Review by ArthurZucker — COMMENTED** [review]
Hey! Great work here, would be great going forward to have an estimatate of what kind of speedup we are expecting (and maybe add it to the readme here) or just benchmark that we indeed have a speedup!

**Comment by susnato** [comment]
> Hey! Great work here, would be great going forward to have an estimatate of what kind of speedup we are expecting (and maybe add it to the readme here) or just benchmark that we indeed have a speedup!

I would love to do it but I don't really have access to a high quality GPU. Also we should keep the GPU same across all benchmarks for all models.

**Comment by younesbelkada** [comment]
@susnato I will run benchmarks for you :)

**Comment by susnato** [comment]
Thanks @younesbelkada ! 
BTW I think you should do it for all models to keep the system configuration constant across all benchmarks.

**Review by younesbelkada — APPROVED** [review]
Thanks again! 
I ran the benchmarks and I am getting 15% - 40% speedup depending on the sequence length, given the fact that it makes training and inference more memory efficient I think this is great to have

Screenshot 2023-10-03 at 13 00 08

Screenshot 2023-10-03 at 13 01 45

File: docs/source/en/perf_infer_gpu_one.md
Comment (ArthurZucker) [inline docs/source/en/perf_infer_gpu_one.md:36]: Same comment as for Llama, let's add a section in the readme of starcoder about expected gains!

**Review by ArthurZucker — APPROVED** [review]
Thanks! If you can update the Readme following [this](https://github.com/huggingface/transformers/pull/26464/files#diff-1338d233bb7a5aeea295120499418f351323de3402888697a1d955811da2d9db) would be great!

**Comment by susnato** [comment]
Thanks for updating the readme of `starcoder`, @younesbelkada!
Hi @ArthurZucker, anything more required to make this ready for merge?

**Comment by younesbelkada** [comment]
Hi @susnato for me the changes look great, I will let @ArthurZucker give a final pass and merge!

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py
Comment (ArthurZucker) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:374]: Let's wait until your PR is merge @younesbelkada 😉

**Review by Ltrack — APPROVED** [review]

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py
Comment (ArthurZucker) [inline src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:235]: I think we should wait again for the padding mask refactor and not pass padding mask! #26792

**Review by ArthurZucker — COMMENTED** [review]
Sorry to block again!

**Comment by cmosguy** [comment]
@younesbelkada @ArthurZucker i was trying to run the blog tutorial code here: https://github.com/pacman100/DHS-LLM-Workshop/blob/main/personal_copilot/training/train.py

I just pulled in the main line and getting this error.

This was working back in August, but now I am getting errors related to the Flash Attention 2.0 error message above in this issue. Why was this working back in August now it is broken? Is there a pinned version if transformers I should be using?

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py
Comment (younesbelkada) [inline src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:303]: There should be no more `padding_mask` now since #26792 has been merged, let me know if you want to handle this, otherwise I can have a look!

Comment (susnato) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:303]: hey I am currently working to fix this.

**Comment by susnato** [comment]
Hello @younesbelkada, I have pushed the commit to remove `padding_mask` and all the tests are passing too!

Screenshot from 2023-10-30 17-02-34

 Please let me know if any more changes are needed.

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py
Comment (younesbelkada) [inline src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:383]: Can you replace that with a `hasattr` logic ? the reason behind it is presented here: https://github.com/huggingface/transformers/pull/26916

**Review by younesbelkada — COMMENTED** [review]
Thanks very much for your great work! LGTM with only 1 nit!

**Review by younesbelkada — APPROVED** [review]

**Comment by susnato** [comment]
Hi @younesbelkada, done!

**Comment by cmosguy** [comment]
@susnato @younesbelkada , sorry to keep bugging you guys here in this PR, but can you please tell me how folks were able to run with Flash Attention back in August using this [tutorial ](https://github.com/pacman100/DHS-LLM-Workshop/blob/main/personal_copilot/training/train.py)form @pacman100 

When I run with the flash attention flag, I keep getting:
```
 File "/opt/lib/python3.10/site-packages/transformers/modeling_utils.py", line 1265, in _check_and_enable_flash_attn_2
        raise ValueError(raise ValueError(

ValueError: The current architecture does not support Flash Attention 2.0. Please open an issue on GitHub to request support for this architecture: https://github.com/huggingface/transformers/issues/newValueError
: The current architecture does not support Flash Attention 2.0. Please open an issue on GitHub to request support for this architecture: https://github.com/huggingface/transformers/issues/new
    raise ValueError(
ValueError: The current architecture does not support Flash Attention 2.0. Please open an issue on GitHub to request support for this architecture: https://github.com/huggingface/transformers/issues/new
    raise ValueError(
ValueError: The current architecture does not support Flash Attention 2.0. Please open an issue on GitHub to request support for this architecture: https://github.com/huggingface/transformers/issues/new
```

**Comment by susnato** [comment]
Hi @cmosguy, could you please try and checkout this branch (`susnato:flash_attn_starcoder`) and re-run your tutorial and let us know if this error is solved or not?

**Comment by cmosguy** [comment]
> Hi @cmosguy, could you please try and checkout this branch (`susnato:flash_attn_starcoder`) then install the library from this branch and re-run your tutorial and let us know if this error is solved or not?
> 
> Right now, StarCoder does not support `flash-attention`, this PR is adds flash attention feature to the model. So you can wait for the PR to get merged or if it is urgent then you can checkout my branch and install from it(like I said above).

I just checked out your branch and ran `pip install -e .` in the transformers library.  After installing I get the following output (sorry for long text):

```text
  File "/opt/../DHS-LLM-Workshop/personal_copilot/training/train.py", line 275, in create_and_prepare_model
        model = AutoModelForCausalLM.from_pretrained(
model = AutoModelForCausalLM.from_pretrained(  File "/opt/ds_research/transformers/src/transformers/models/auto/auto_factory.py", line 566, in from_pretrained

  File "/opt/ds_research/transformers/src/transformers/models/auto/auto_factory.py", line 566, in from_pretrained
    model = AutoModelForCausalLM.from_pretrained(
  File "/opt/ds_research/transformers/src/transformers/models/auto/auto_factory.py", line 566, in from_pretrained
    model = AutoModelForCausalLM.from_pretrained(
  File "/opt/ds_research/transformers/src/transformers/models/auto/auto_factory.py", line 566, in from_pretrained
        return model_class.from_pretrained(
  File "/opt/ds_research/transformers/src/transformers/modeling_utils.py", line 3175, in from_pretrained
return model_class.from_pretrained(
  File "/opt/ds_research/transformers/src/transformers/modeling_utils.py", line 3175, in from_pretrained
    return model_class.from_pretrained(
  File "/opt/ds_research/transformers/src/transformers/modeling_utils.py", line 3175, in from_pretrained
    return model_class.from_pretrained(
  File "/opt/ds_research/transformers/src/transformers/modeling_utils.py", line 3175, in from_pretrained
    config = cls._check_and_enable_flash_attn_2(config, torch_dtype=torch_dtype, device_map=device_map)
  File "/opt/ds_research/transformers/src/transformers/modeling_utils.py", line 1275, in _check_and_enable_flash_attn_2
        config = cls._check_and_enable_flash_attn_2(config, torch_dtype=torch_dtype, device_map=device_map)config = cls._check_and_enable_flash_attn_2(config, torch_dtype=torch_dtype, device_map=device_map)
  File "/opt/ds_research/transformers/src/transformers/modeling_utils.py", line 1275, in _check_and_enable_flash_attn_2

  File "/opt/ds_research/transformers/src/transformers/modeling_utils.py", line 1275, in _check_and_enable_flash_attn_2
    config = cls._check_and_enable_flash_attn_2(config, torch_dtype=torch_dtype, device_map=device_map)
  File "/opt/ds_research/transformers/src/transformers/modeling_utils.py", line 1275, in _check_and_enable_flash_attn_2
    raise ImportError(
ImportError: Flash Attention 2 is not available. Please refer to the documentation of https://github.com/Dao-AILab/flash-attention for installing it. Make sure to have at least the version 2.1.0
            raise ImportError(raise ImportError(

raise ImportError(
ImportErrorImportErrorImportError: : : Flash Attention 2 is not available. Please refer to the documentation of https://github.com/Dao-AILab/flash-attention for installing it. Make sure to have at least the version 2.1.0Flash Attention 2 is not available. Please refer to the documentation of https://github.com/Dao-AILab/flash-attention for installing it. Make sure to have at least the version 2.1.0Flash Attention 2 is not available. Please refer to the documentation of https://github.com/Dao-AILab/flash-attention for installing it. Make sure to have at least the version 2.1.0


[2023-10-30 07:55:00,169] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 0 (pid: 42849) of binary: /opt/bin/python
Traceback (most recent call last):
  File "/opt/bin/accelerate", line 8, in <module>
    sys.exit(main())
  File "/opt/lib/python3.10/site-packages/accelerate/commands/accelerate_cli.py", line 47, in main
    args.func(args)
  File "/opt/lib/python3.10/site-packages/accelerate/commands/launch.py", line 985, in launch_command
    multi_gpu_launcher(args)
  File "/opt/lib/python3.10/site-packages/accelerate/commands/launch.py", line 654, in multi_gpu_launcher
    distrib_run.run(args)
  File "/opt/lib/python3.10/site-packages/torch/distributed/run.py", line 797, in run
    elastic_launch(
  File "/opt/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/opt/lib/python3.10/site-packages/torch/distributed/launcher/api.py", line 264, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
/opt/../DHS-LLM-Workshop/personal_copilot/training/train.py FAILED
------------------------------------------------------------
```

**Comment by susnato** [comment]
Hi @cmosguy, could you please run `pip install flash-attn --no-build-isolation` and then re-run the script? 
Or check the flash attention version on your system? `pip freeze | grep flash`  ?

**Comment by cmosguy** [comment]
Hey @susnato -

I had no idea that I had to install that additonal package.  Shouldn't that have been installed with the `setup.py` from `transformers` library?

Anyways, I repeated the commands you mentioned:
```bash
pip freeze | grep flash
flash-attn==2.3.3
```

So it looks like it is loading:
```text
You are attempting to use Flash Attention 2.0 without specifying a torch dtype. This might lead to unexpected behaviour
You are attempting to use Flash Attention 2.0 with a model initialized on CPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.
You are attempting to use Flash Attention 2.0 without specifying a torch dtype. This might lead to unexpected behaviour
You are attempting to use Flash Attention 2.0 without specifying a torch dtype. This might lead to unexpected behaviour
You are attempting to use Flash Attention 2.0 with a model initialized on CPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.
You are attempting to use Flash Attention 2.0 with a model initialized on CPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.
You are attempting to use Flash Attention 2.0 without specifying a torch dtype. This might lead to unexpected behaviour
You are attempting to use Flash Attention 2.0 with a model initialized on CPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.
```
I am assuming at this point the `flash-attn` is kicking in.  

However, I am getting the following errors with `OOM`.  I am not sure if this is related or what I can do differently:

```text
File "/opt/lib/python3.10/site-packages/accelerate/accelerator.py", line 1433, in prepare_model
    model = torch.nn.parallel.DistributedDataParallel(
  File "/opt/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 809, in __init__
    self._ddp_init_helper(
  File "/opt/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1098, in _ddp_init_helper
    self.reducer = dist.Reducer(
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 57.81 GiB. GPU 1 has a total capacty of 79.35 GiB of which 19.56 GiB is free. Including non-PyTorch memory, this process has 59.78 GiB memory in use. Of the allocated memory 57.87 GiB is allocated by PyTorch, and 1.60 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```
I've been struggling with this even when I do disable `flash-attention`.  I do not mean to hijack this PR with this information though.  If you do have some ideas how to work around this it would be awesome! If not that's ok too :)

**Comment by susnato** [comment]
Hey @cmosguy, I will checkout the script Wednesday and get back to you if I find a solution :)

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py
Comment (amyeroberts) [inline src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:245]: Let's use `logger.warning_once` here so that the user can control the verbosity and only see's it on the first forward call

**Review by amyeroberts — APPROVED** [review]
Thanks for all of the work adding this! 

Just a few small nits

File: docs/source/en/perf_infer_gpu_one.md
Comment (susnato) [reply docs/source/en/perf_infer_gpu_one.md:37]: Hey, if I remember it correctly, it was asked by @ArthurZucker

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py
Comment (susnato) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:318]: Ok

File: docs/source/en/perf_infer_gpu_one.md
Comment (amyeroberts) [reply docs/source/en/perf_infer_gpu_one.md:37]: OK - it doesn't really matter so I'm happy to leave it and find out when Arthur's back what that reasoning is :)

Comment (susnato) [reply docs/source/en/perf_infer_gpu_one.md:37]: Ok then leaving it as it is.

File: docs/source/en/model_doc/gpt_bigcode.md
Comment (susnato) [reply docs/source/en/model_doc/gpt_bigcode.md:72]: currently trying to get the output and then I will push the changes.

Comment (susnato) [reply docs/source/en/model_doc/gpt_bigcode.md]: Done.

**Comment by susnato** [comment]
Hi @amyeroberts, I have pushed the changes.

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py
Comment (sohamparikh94) [inline src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:367]: hi @susnato, sorry for commenting on an already merged PR. Should this not be `self.config.attn_pdrop` instead of `self.dropout`?

Comment (sohamparikh94) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:367]: sure, i'd be happy to help. I can do that after a few hours if that's okay (have work commitments + it's my first time contributing)

Comment (susnato) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:367]: Hi @sohamparikh94, I am sorry, I already did open a [PR](https://github.com/huggingface/transformers/pull/27182#pullrequestreview-1706573490).

Comment (sohamparikh94) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:367]: no problem, thanks a lot!

**Comment by susnato** [comment]
Hello @cmosguy, the dropout error is fixed now on the main branch.

Could you please install again from the main branch and re-run your script? It should work now.

**Comment by susnato** [comment]
Hi @cmosguy, I ran the script with some minor hyper-parameter changes (to suit my GPU) and it's working!


Screenshot from 2023-11-02 04-25-17
Screenshot from 2023-11-02 04-25-42

Make sure to override (`use_flash_attention_2=True`) in this [line](https://github.com/pacman100/DHS-LLM-Workshop/blob/79a978e204393f4c39df8553afd5c5a2aa43ba9c/personal_copilot/training/train.py#L282) if you feel that Flash Attention is not being used here.

Also don't forget to re-install from the main :).

**Comment by cmosguy** [comment]
Hey @susnato Thanks for your efforts here, yes I was able to see the training start with the lines:
```
You are attempting to use Flash Attention 2.0 without specifying a torch dtype. This might lead to unexpected behaviour
You are attempting to use Flash Attention 2.0 with a model initialized on CPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.
```
Thanks for your help here.

If you do not mind me asking, what happened back in August?  The tutorial I mentioned with script had things working before, but you made a lot of edits that indicates this is just now being added. Was it there before then removed?  I guess I am  trying to understand, because I may be interested in swapping in other models and I cannot fully comprehend when flash attention can be used or not on which model.

**Comment by susnato** [comment]
Hello @cmosguy, I am glad that you were able to start training :).

> If you do not mind me asking, what happened back in August? The tutorial I mentioned with script had things working before, but you made a lot of edits that indicates this is just now being added. Was it there before then removed?

As far as I know Flash Attention is just added for this model. 
Actually if we look at this [commit](https://github.com/pacman100/DHS-LLM-Workshop/commit/f4693e707ae82f9a7124d5d9a549daa827be7169) from the same tutorial you provided, back in August whenever you would load any of the `Falcon`, `llama` or `starcoder` models, your model's attention forward code would be replaced by the custom flash attention code so you would be able to use Flash Attention without any hesitation.([replace_starcoder_attn_with_flash_attn](https://github.com/pacman100/DHS-LLM-Workshop/blob/main/personal_copilot/training/starcoder_flash_attn_monkey_patch.py), [replace_llama_attn_with_flash_attn](https://github.com/pacman100/DHS-LLM-Workshop/blob/main/personal_copilot/training/llama_flash_attn_monkey_patch.py), [replace_falcon_attn_with_flash_attn](https://github.com/pacman100/DHS-LLM-Workshop/blob/main/personal_copilot/training/falcon_flash_attn_monkey_patch.py)).

But then Flash Attention were added to `Flacon` and `Llama` models (in `transformer` main), and @pacman100 removed this block which used to modify the attention code. It was okay for `falcon` and `llama` but since `starcoder` didn't get that feature yet, you were getting the errors whenever you tried to use `starcoder`. But now `starcoder` has FlashAttention so it's all good.

Also please note that the flash attention code from the tutorial didn't handle `attention_mask` but it is properly handled by the `transformers` main.  

I hope this explanation helps, otherwise feel free to tag me if you need any help. :)

> I may be interested in swapping in other models and I cannot fully comprehend when flash attention can be used or not on which model.

For every model that supports `Flash Attention`, will have `_supports_flash_attn_2` as True. You can just load the model and check that, for example - 

```python
from transformers import AutoModel

model = AutoModel.from_pretrained("bigcode/starcoderbase-1b")
model._supports_flash_attn_2 # will give output as True since it has support for Flash Attention
```

```python
from transformers import AutoModel

model = AutoModel.from_pretrained("gpt2")
model._supports_flash_attn_2 # will give output as False since it does not has support for Flash Attention (For now).
```

**Comment by cmosguy** [comment]
@susnato  wow I am learning so much hanging out here.  Thank you for walking me through what happened this really explains a lot.  I appreciate you taking the time to investigate the issue and writing a coherent explanation.  I totally did not see the [commit](https://github.com/pacman100/DHS-LLM-Workshop/commit/f4693e707ae82f9a7124d5d9a549daa827be7169) from before, thank you for bringing this to my attention (pun intended).  OK, so I am off to the races here with the training in the meantime, cheers!

**Comment by susnato** [comment]
Hey @cmosguy, it was fun for me too! 

cheers!

**Pull request merged**
